packages/iota/src/validation/payloads/payloads.ts

```diff
@@ -55,16 +60,36 @@ export function validatePayload(
 export function validateTransactionPayload(
     transactionPayload: ITransactionPayload, protocolInfo: INodeInfoProtocol
 ): IValidationResult {
-    let txEssenceResult: IValidationResult = { isValid: true };
-    let unlocksResult: IValidationResult = { isValid: true };
+    const results: IValidationResult[] = [];
 
     if (transactionPayload.type === TRANSACTION_PAYLOAD_TYPE) {
-        txEssenceResult = validateTransactionEssence(transactionPayload.essence, protocolInfo);
-        unlocksResult = validateUnlocks(transactionPayload.unlocks);
+        const unlocksCount = transactionPayload.unlocks.length;
+        const inputsCount = transactionPayload.essence.inputs.length;
+
+        if (unlocksCount !== inputsCount) {
+            results.push({
+                isValid: false,
+                errors: ["Transaction payload unlocks count must match inputs count of the Transaction Essence."]
+            });
+        }
+        results.push(validateTransactionEssence(transactionPayload.essence, protocolInfo));
+        results.push(validateUnlocks(transactionPayload.unlocks));
     } else {
         throw new Error(`Unrecognized transaction type ${transactionPayload.type}`);
     }
 
-    return mergeValidationResults(txEssenceResult, unlocksResult);
+    return mergeValidationResults(...results);
 }
 
+/**
+ * Validate a tagged data payload.
+ * @param taggedDataPayload The tagged data payload to validate.
+ * @returns The validation result.
+ */
+export function validateTaggedDataPayload(
+    taggedDataPayload: ITaggedDataPayload
+): IValidationResult {
+    return (taggedDataPayload.tag && taggedDataPayload.tag.length / 2 > MAX_TAG_LENGTH) ?
+        failValidation({ isValid: true }, `Tagged Data Payload tag length exceeds the maximum size of ${MAX_TAG_LENGTH}.`) :
```

Comment:
This looks a bit weird, `failValidation` and the `isValid: true`. Just make it `false`.